src/libraries/System.Private.CoreLib/src/System/Enum.cs

```diff
@@ -258,11 +258,27 @@ internal static ulong ToUInt64(object value)
             return result;
         }
 
+        private static ulong ToUInt64<TEnum>(TEnum value) where TEnum : struct, Enum =>
+            Type.GetTypeCode(typeof(TEnum)) switch
+            {
+                TypeCode.SByte => (ulong)Unsafe.As<TEnum, sbyte>(ref value),
+                TypeCode.Byte => Unsafe.As<TEnum, byte>(ref value),
+                TypeCode.Boolean => Convert.ToByte(Unsafe.As<TEnum, bool>(ref value)),
+                TypeCode.Int16 => (ulong)Unsafe.As<TEnum, short>(ref value),
+                TypeCode.UInt16 => Unsafe.As<TEnum, ushort>(ref value),
+                TypeCode.Char => Unsafe.As<TEnum, char>(ref value),
+                TypeCode.UInt32 => Unsafe.As<TEnum, uint>(ref value),
+                TypeCode.Int32 => (ulong)Unsafe.As<TEnum, int>(ref value),
+                TypeCode.UInt64 => Unsafe.As<TEnum, ulong>(ref value),
+                TypeCode.Int64 => (ulong)Unsafe.As<TEnum, long>(ref value),
+                _ => throw new InvalidOperationException(SR.InvalidOperation_UnknownEnumType),
+            };
```

Comment:
Wonder if this impl is better:
```csharp
private static ulong ToUInt64<TEnum>(TEnum value) where TEnum : struct, Enum
{
    if (!typeof(TEnum).IsPrimitive) // is not currently optimized in JIT but can be.
        throw new InvalidOperationException(SR.InvalidOperation_UnknownEnumType);
    return Unsafe.SizeOf<TEnum>() switch
    {
        8 => Unsafe.As<TEnum, ulong>(ref value),
        4 => Unsafe.As<TEnum, uint>(ref value),
        2 => Unsafe.As<TEnum, ushort>(ref value),
        _ => Unsafe.As<TEnum, byte>(ref value),
    };
}
```

Reply:
Also, I recall it's actually possible to create a float/double based Enum (and perhaps IntPtr/UIntPtr?)
at least we had a test for it somewhere.

Reply:
> Also, I recall it's actually possible to create a float/double based Enum (and perhaps IntPtr/UIntPtr?)
at least we had a test for it somewhere.

That already wasn't supported:
https://github.com/dotnet/runtime/blob/640f0f72bc1c31717b6a319ced7e00bd44675349/src/libraries/System.Private.CoreLib/src/System/Enum.cs#L256

> Wonder if this impl is better

IsPrimitive returns true for float/double.

Reply:
I'll take a look if `Type.GetTypeCode` can be optimized for `typeof()` input into a const.